### PR TITLE
feat(regional): Bank remittance report

### PR DESCRIPTION
**Pull request by Mangesh-Khairnar**
- Allow creation of payment order based on payment entry
creating_payment_entry_from_payment_order

- Add a generate txt report for bank remittance for Indian clients and this will generate and attach the text file to the doc and also, initiate download.
Screenshot 2019-05-08 at 4 30 03 PM

- Text report generated according to the multiband format followed by banks
Screenshot 2019-05-09 at 4 53 52 PM

- custom fields in company added for india specific client
Screenshot 2019-05-06 at 4 16 27 PM

### CODE REVIEW

File: erpnext/regional/india/bank_remittance_txt.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		}, 'parent')
+
+	supplier_billing_address = frappe.get_cached_doc('Address', addr_link)
+	email = ','.join(filter(None, [supplier_billing_address.email_id, company_email]))
+
+	detail = OrderedDict(
+		record_identifier='D',
+		payment_ref_no=sanitize_data(ref_doc.payment_entry),
+		payment_type=cstr(payment_entry.mode_of_payment)[:10],
+		amount=str(format(ref_doc.amount, '.2f'))[:13],
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: Dont clip amount

File: erpnext/regional/india/bank_remittance_txt.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+from frappe.model.document import Document
+from frappe.utils import cint,cstr, today
+import re
+import datetime
+from collections import OrderedDict
+from frappe.core.doctype.file.file import download_file
+
+def create_bank_remittance_txt(name):
+	payment_order = frappe.get_doc("Payment Order", name)
+
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: validate mandatory values from payment order

**Review by netchampfaris — CHANGES_REQUESTED** [review]

File: erpnext/regional/india/bank_remittance_txt.py (hunk 1 above)
Comment (Mangesh-Khairnar) [reply erpnext/regional/india/bank_remittance_txt.py]: fixed

File: erpnext/regional/india/bank_remittance_txt.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+def format_date(val):
+	''' Convert a datetime object to DD/MM/YYYY format '''
+	return val.strftime("%d/%m/%Y")
+
+def validate_amount(val, max_int_size):
+	''' Validate amount to be within the allowed limits  '''
+	int_size = len(str(val).split('.')[0])
+
+	if int_size > max_int_size:
+		frappe.throw(_("Amount for a single transaction is more than maximum allowed amount, create a separate payment order by splitting the transactions"))
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: Specify how much is the maximum amount in the message

File: erpnext/regional/india/bank_remittance_txt.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+	return val
+
+def validate_information(obj, attr, max_size):
+	''' Checks if the information is not set in the system and is within the size '''
+	if hasattr(obj, attr):
+		return validate_field_size(getattr(obj, attr), frappe.unscrub(attr), max_size)
+
+	else:
+		frappe.throw(_("%s is mandatory for generating remittance payments, set the field and try again" % attr))
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: Use {0} placeholder in translated strings

File: erpnext/regional/india/bank_remittance_txt.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+def validate_field_size(val, label, max_size):
+	''' check the size of the val '''
+	if len(cstr(val)) > max_size:
+		frappe.throw("%s field is limited to size %d" % (label, max_size))
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: Add translated string

File: erpnext/regional/india/bank_remittance_txt.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	return "~".join(header), file_name
+
+def get_batch_row(doc, no_of_records, total_amount, product_code):
+	batch = ["B"]
+	batch.append(validate_field_size(no_of_records, "No Of Records", 5))
+	batch.append(validate_amount(format(total_amount, '0.2f'), 17))
+	batch.append(sanitize_data(doc.name, '_')[:20])
+	batch.append(format_date(doc.posting_date))
+	batch.append(validate_field_size(product_code,"Product Code", 20))
+	print(batch)
```
Comment (netchampfaris) [inline erpnext/regional/india/bank_remittance_txt.py]: Remove print

**Review by netchampfaris — CHANGES_REQUESTED** [review]

**Review by netchampfaris — APPROVED** [review]

**Pull request merged**